source/common/http/http2/codec_impl.cc

```diff
@@ -1107,104 +1176,50 @@ Status ConnectionImpl::onFrameReceived(const nghttp2_frame* frame) {
   // and CONTINUATION frames in onBeforeFrameReceived().
   ASSERT(frame->hd.type != NGHTTP2_CONTINUATION);
 
-  if ((frame->hd.type == NGHTTP2_PING) && (frame->ping.hd.flags & NGHTTP2_FLAG_ACK)) {
+  if (frame->hd.type == NGHTTP2_PING) {
     // The ``opaque_data`` should be exactly what was sent in the ping, which is
     // was the current time when the ping was sent. This can be useful while debugging
     // to match the ping and ack.
     uint64_t data;
     safeMemcpy(&data, &(frame->ping.opaque_data));
-    ENVOY_CONN_LOG(trace, "recv PING ACK {}", connection_, data);
-
-    onKeepaliveResponse();
-    return okStatus();
+    return onPing(data, frame->ping.hd.flags & NGHTTP2_FLAG_ACK);
   }
 
   if (frame->hd.type == NGHTTP2_DATA) {
-    RETURN_IF_ERROR(trackInboundFrames(frame->hd.stream_id, frame->hd.length, frame->hd.type,
-                                       frame->hd.flags, frame->data.padlen));
+    return onBeginData(frame->hd.stream_id, frame->hd.length, frame->hd.type, frame->hd.flags,
+                       frame->data.padlen);
   }
 
   // Only raise GOAWAY once, since we don't currently expose stream information. Shutdown
   // notifications are the same as a normal GOAWAY.
   // TODO: handle multiple GOAWAY frames.
-  if (frame->hd.type == NGHTTP2_GOAWAY && !raised_goaway_) {
-    ASSERT(frame->hd.stream_id == 0);
```

Comment:
optional: WDYT of keeping the assert either here or in onGoAway()? (I'm not sure why it was added.)

Reply:
Sure, I'll keep it here for now. In a future PR, I will be removing the onFrameReceived() method in favor of Http2VisitorInterface method implementations. The GOAWAY visitor method in that case does not provide the frame header's stream ID.